### PR TITLE
fix(asm): fix cpu leak

### PR DESCRIPTION
**Pull request by christophe-papazian**
This PR fixes a cpu leak in ASM instrumentation, that would add additional overhead to all requests if ASM, API Security and Telemetry were enabled on Flask instrumentation.

Related to incident 25768.

Add one regression unit test for all ASM supported frameworks.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/appsec/_asm_request_context.py [hunk 1]
```diff
@@ -151,7 +151,7 @@ def finalise(self):
             callbacks = GLOBAL_CALLBACKS.get(_CONTEXT_CALL, []) if env.must_call_globals else []
             env.must_call_globals = False
             if env is not None and env.callbacks is not None and env.callbacks.get(_CONTEXT_CALL):
```
Comment (P403n1x87) [inline ddtrace/appsec/_asm_request_context.py:153]: `env.callbacks.get(_CONTEXT_CALL)` is being called twice. Perhaps the first call can be assigned and re-used?

Comment (christophe-papazian) [reply ddtrace/appsec/_asm_request_context.py:153]: it's a get on a native dict, so very fast. I prefer to not introduce performance improvement in this PR, I opened another PR for that to be merged later: https://github.com/DataDog/dd-trace-py/pull/8617
This is only bug fix.

**Review by emmettbutler — APPROVED** [review]
This is up there with avoiding the use of mutable keyword arguments on the list of Python toe-stubbers
https://stackoverflow.com/a/2347423/735204

File: ddtrace/appsec/_asm_request_context.py [hunk 2]
```diff
@@ -151,7 +151,7 @@ def finalise(self):
...
-                callbacks += env.callbacks.get(_CONTEXT_CALL)
+                callbacks = callbacks + env.callbacks.get(_CONTEXT_CALL)
```
Comment (Julio-Guerra) [inline ddtrace/appsec/_asm_request_context.py:154]: It's worth mentioning for future devs going through this code why `+=` isn't equivalent.

Comment (Julio-Guerra) [reply ddtrace/appsec/_asm_request_context.py:154]: Quoting [this SO answer](https://stackoverflow.com/questions/2347265/why-does-behave-unexpectedly-on-lists/2347423#2347423):
> For types that support both __iadd__ and __add__ you therefore have to be careful which one you use. a += b will call __iadd__ and mutate a, whereas a = a + b will create a new object and assign it to a. They are not the same operation!

Comment (Julio-Guerra) [reply ddtrace/appsec/_asm_request_context.py:154]: I also wonder if static analyzer could help warn here about an unintended side effect.

**Review by gnufede — APPROVED** [review]

**Pull request merged**